cdap-formats/src/main/java/io/cdap/cdap/format/io/StructuredRecordDatumWriter.java

```diff
@@ -67,6 +69,20 @@ public void encode(StructuredRecord data, Encoder encoder) throws IOException {
    * @throws IOException If failed to encode
    */
   protected void encode(Encoder encoder, Schema schema, Object value) throws IOException {
+    //handle any logical types that need special handling
+    if (schema.getLogicalType() != null) {
```

Comment:
I don't think we need to handle logical types at this level. Encoding should just use whatever physical type is there.

Enforcing schema correctness is its own piece of work that should be done at the StructuredRecord level, to ensure it isn't possible to create an invalid record (CDAP-15317).